engine/chunker/cached_chunker.go

```diff
@@ -0,0 +1,516 @@
+		err = ls.performOnCache(ctx, func(cache *lru.Cache) { cache.Add(l, links) })
+		if err != nil {
+			return err
+		}
+		count++
+	}
+
+	// If no root key is present in datastore, it means the cache should be empty
+	// Therefore, clear all keys in the datastore.
+	//
+	// This also makes sure that data cached using previous implementation of caching is cleared.
+	if count == 0 {
+		// Query all keys in datastore.
+		allKeys := dsq.Query{
+			KeysOnly: true,
+		}
+		allKeysResult, err := ls.ds.Query(ctx, allKeys)
+		if err != nil {
+			return err
+		}
+		defer allKeysResult.Close()
+
+		// For each key in datastore delete its corresponding value.
+		var prunedCount int
+		for r := range allKeysResult.Next() {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			if r.Error != nil {
+				return fmt.Errorf("cannot read cache key: %w", r.Error)
+			}
+			err := ls.ds.Delete(ctx, datastore.RawKey(r.Key))
+			if err != nil {
+				return err
+			}
+			prunedCount++
+		}
+
+		// If datastore was pruned, log informative message.
+		if prunedCount != 0 {
+			log.Infow("No caching metadata is persisted but datastore is non-empty; pruned lingering cache entries", "count", prunedCount)
+		}
+	} else {
+		log.Debugw("Cache restored successfully", "restoredCacheCount", ls.Len(), "capacity", ls.Cap())
+	}
+
+	// If the cache capacity was too small to restore all entries present, it means cache was
+	// evicted during restore and records were pruned as needed.
+	//
+	// Log an informative message to let the user know.
+	if ls.Cap() < count {
+		log.Infow("Cache capacity is smaller than previously persisted cache; pruned persisted cache.", "persistedCacheCount", count, "capacity", ls.cache.MaxEntries)
+	}
+	return nil
+}
+
+// performOnCache is a utility to perform operatons in CachedEntriesChunker.cache to safely set
+// the context to be used during eviction and return errors that may occur as a result of
+// eviction if performing the given action indeed causes it.
+func (ls *CachedEntriesChunker) performOnCache(ctx context.Context, action func(*lru.Cache)) error {
+	ls.onEvictedCtx = ctx
+	defer func() {
+		ls.onEvictedCtx = nil
+		ls.onEvictedErr = nil
+	}()
+	action(ls.cache)
+	err := ls.onEvictedErr
+	return err
+}
+
+// listEntriesChainLinks lists the links to the entries chain with given root.
+// The returned list will always include the root itself.
+//
+// Note that if traversal of the chain partially fails any links listed so far will be returned
+// along with the error.
+func (ls *CachedEntriesChunker) listEntriesChainLinks(ctx context.Context, root ipld.Link) ([]ipld.Link, error) {
+	var links []ipld.Link
+	lCtx := ipld.LinkContext{Ctx: ctx}
+	next := root
+	for {
+		n, err := ls.lsys.Load(lCtx, next, schema.Type.EntryChunk)
+		if err != nil {
+			return links, err
+		}
+		chunk, ok := n.(schema.EntryChunk)
+		if !ok {
+			return links, fmt.Errorf("expected schema.EntryChunk for link: %s", next)
```

Comment:
```suggestion
			return links, fmt.Errorf("expected schema.EntryChunk for link: %s", next.(cidlink.Link).Cid)
```
Do you need to convert the `next` link to a CID to get that string in the error message?

Reply:
I don't think we need to, because `ipld.Link`[ implements `Stringer` interface](https://github.com/ipld/go-ipld-prime/blob/master/datamodel/link.go#L33-L36).

The change suggested above:
1.  is equivalent to the code that is there already, since `cidlink.Link.String()` simply calls `cid.String()` [here](https://github.com/ipld/go-ipld-prime/blob/master/linking/cid/cidLink.go#L30-L32), and
2. assumes `next` is of type `cidlink.Link`; sure that is true today but if IPLD makes ca change to that for whatever reason we end up getting a panic in such lines. 


Try running this:
```go
func Test_StringerErrorMessage(t *testing.T) {
	c, err := cid.Decode("baguqeeqq4dalrhkuucrvzkmy6louau2wm4")
	if err != nil {
		t.Fatal(err)
	}
	l := cidlink.Link{Cid: c}

	errFromCid := fmt.Errorf("%s", c)
	errFromLink := fmt.Errorf("%s", l)

	if errFromCid.Error() != errFromLink.Error() {
		t.Fatal("andrew is right")
	}
}
```
 🙂